### PR TITLE
domain: record latest schema version after syncer restarted

### LINKED ISSUE #49773: admin check failed with “Error 8223 (HY000): data inconsistency in table” after add index and inject ddl owner network partition
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、tidb_enable_dist_task='on'
2、run sysbench
3、add index for one of table
4、inject network partition between ddl owner and all other pods
5、admin check index

### 2. What did you expect to see? (Required)
admin check can success

### 3. What did you see instead (Required)
admin check failed：
admin check index sbtest1 index_test_1703552562599
Error 8223 (HY000): data inconsistency in table: sbtest1, index: index_test_1703552562599, handle: 31702184, index-values:"handle: 31702184, values: [KindString 42604366371-21451895812-27518569128-37799479491-91566181356-76146461753-53016366997-52301770304-09657853519-55786043329]" != record-values:"handle: 31702184, values: [KindString 96558301162-97928307451-01675937654-31540943557-87200715168-69587168424-27210814037-31751409618-31353104576-34317501676]"

operatorLogs:
[2023-12-26 09:02:42] ###### start adding index
alter table sbtest1 add index index_test_1703552562599 (c)
[2023-12-26 09:02:42] ###### wait for ddl job finish
[2023-12-26 09:06:43] ###### ddl job finished
select job_id, job_type, state from information_schema.ddl_jobs where query = 'alter table sbtest1 add index index_test_1703552562599 (c)'
jobId: 561, job type: add index /* ingest */, state: synced
add index done, it takes: 4m0.473890156s
[2023-12-26 09:06:43] ###### start admin check
admin check index sbtest1 index_test_1703552562599

tidb logs：
[endless-ha-test-add-index-tps-5251454-1-745.zip](https://github.com/pingcap/tidb/files/13776182/endless-ha-test-add-index-tps-5251454-1-745.zip)

[2023/12/26 09:11:24.460 +08:00] [ERROR] [reporter.go:291] ["admin check found data inconsistency"] [conn=3877634872] [session_alias=] [table_name=sbtest1] [index_name=index_test_1703552562599] [row_id=31702184] [index="handle: 31702184, values: [KindString 42604366371-21451895812-27518569128-37799479491-91566181356-76146461753-53016366997-52301770304-09657853519-55786043329]"] [row="handle: 31702184, values: [KindString 96558301162-97928307451-01675937654-31540943557-87200715168-69587168424-27210814037-31751409618-31353104576-34317501676]"] [row_mvcc="{\"decoded\":{\"429391636255211523\":{\"c\":\"42604366371-21451895812-27518569128-37799479491-91566181356-76146461753-53016366997-52301770304-09657853519-55786043329\",\"k\":\"35159807\",\"pad\":\"21788622147-59129949870-21609172825-73750306916-68396718180\"},\"446576145727488168\":{\"c\":\"96558301162-97928307451-01675937654-31540943557-87200715168-69587168424-27210814037-31751409618-31353104576-34317501676\",\"k\":\"35159807\",\"pad\":\"21788622147-59129949870-21609172825-73750306916-68396718180\"}},\"key\":\"7480000000000000F55F728000000001E3BCA8\",\"mvcc\":{\"info\":{\"writes\":[{\"start_ts\":446576145727488168,\"commit_ts\":446576145858560101,\"short_value\":\"gAADAAAAAgMEBAB7ALYA/34YAjk2NTU4MzAxMTYyLTk3OTI4MzA3NDUxLTAxNjc1OTM3NjU0LTMxNTQwOTQzNTU3LTg3MjAwNzE1MTY4LTY5NTg3MTY4NDI0LTI3MjEwODE0MDM3LTMxNzUxNDA5NjE4LTMxMzUzMTA0NTc2LTM0MzE3NTAxNjc2MjE3ODg2MjIxNDctNTkxMjk5NDk4NzAtMjE2MDkxNzI4MjUtNzM3NTAzMDY5MTYtNjgzOTY3MTgxODA=\"},{\"start_ts\":429391636255211523,\"commit_ts\":429391636884094978,\"short_value\":\"gAADAAAAAgMEBAB7ALYA/34YAjQyNjA0MzY2MzcxLTIxNDUxODk1ODEyLTI3NTE4NTY5MTI4LTM3Nzk5NDc5NDkxLTkxNTY2MTgxMzU2LTc2MTQ2NDYxNzUzLTUzMDE2MzY2OTk3LTUyMzAxNzcwMzA0LTA5NjU3ODUzNTE5LTU1Nzg2MDQzMzI5MjE3ODg2MjIxNDctNTkxMjk5NDk4NzAtMjE2MDkxNzI4MjUtNzM3NTAzMDY5MTYtNjgzOTY3MTgxODA=\"}]}},\"regionID\":44171}"] [index_mvcc="{\"decoded\":{\"446576083376537788\":{\"handle\":\"31702184\"}},\"key\":\"7480000000000000F55F69800000000000004F013432363034333636FF3337312D32313435FF313839353831322DFF3237353138353639FF3132382D33373739FF393437393439312DFF3931353636313831FF3335362D37363134FF363436313735332DFF3533303136333636FF3939372D35323330FF313737303330342DFF3039363537383533FF3531392D35353738FF3630343333323900FE038000000001E3BCA8\",\"mvcc\":{\"info\":{\"writes\":[{\"start_ts\":446576083376537788,\"commit_ts\":446576083376537788,\"short_value\":\"MA==\"}]}},\"regionID\":138168}"] [stack="github.com/pingcap/tidb/pkg/util/logutil/consistency.(*Reporter).ReportAdminCheckInconsistent\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/logutil/consistency/reporter.go:290\ngithub.com/pingcap/tidb/pkg/executor.(*checkIndexWorker).HandleTask\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/executor.go:2685\ngithub.com/pingcap/tidb/pkg/resourcemanager/pool/workerpool.(*WorkerPool[...]).handleTaskWithRecover\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/resourcemanager/pool/workerpool/workerpool.go:131\ngithub.com/pingcap/tidb/pkg/resourcemanager/pool/workerpool.(*WorkerPool[...]).runAWorker.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/resourcemanager/pool/workerpool/workerpool.go:147\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/wait_group_wrapper.go:157"]

### 4. What is your TiDB version? (Required)
./tidb-server -V
 Release Version: v7.6.0-alpha
Edition: Community
Git Commit Hash: 07e9ece9107536fad2d5df7664809350f0ce0a4a
Git Branch: heads/refs/tags/v7.6.0-alpha
UTC Build Time: 2023-12-25 11:44:42
GoVersion: go1.21.5
Race Enabled: false
Check Table Before Drop: false
Store: unistore
2023-12-26T08:44:32.225+0800

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #49773

Problem Summary:

```
// tidb-0
[2024/01/02 09:38:06.869 +08:00] [INFO] [ddl.go:1063] ["start DDL job"] [category=ddl] [job="ID:415, Type:add index, State:queueing, SchemaState:none, SchemaID:102, TableID:245, RowCount:0, ArgLen:6, start time: 2024-01-02 09:38:06.845 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0, UniqueWarnings:0"] [query="alter table sbtest1 add index index_test_1704159486844 (c)"]
[2024/01/02 09:38:07.145 +08:00] [INFO] [ddl_worker.go:1030] ["run DDL job"] [worker="worker 6, tp add index"] [category=ddl] [jobID=415] [conn=3644851860] [category=ddl] [job="ID:415, Type:add index, State:running, SchemaState:write only, SchemaID:102, TableID:245, RowCount:0, ArgLen:0, start time: 2024-01-02 09:38:06.845 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0, UniqueWarnings:0"]
[2024/01/02 09:38:07.200 +08:00] [INFO] [domain.go:873] ["mdl gets lock, update to owner"] [jobID=415] [version=55]
...
[2024/01/02 09:39:02.947 +08:00] [INFO] [ddl_worker.go:1030] ["run DDL job"] [worker="worker 5, tp add index"] [category=ddl] [jobID=415] [conn=3644851860] [category=ddl] [job="ID:415, Type:add index, State:running, SchemaState:write reorganization, SchemaID:102, TableID:245, RowCount:0, ArgLen:0, start time: 2024-01-02 09:38:06.845 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:446735184611246206, UniqueWarnings:0"]
[2024/01/02 09:41:57.281 +08:00] [WARN] [repeatable_read.go:282] ["UpdateForUpdateTS failed"] [conn=3644851824] [session_alias=] [error="[tikv:9001]PD server timeout: "]
[2024/01/02 09:41:57.281 +08:00] [INFO] [conn.go:1132] ["command dispatched failed"] [conn=3644851824] [session_alias=] [connInfo="id:3644851824, addr:10.233.71.65:48912 status:11, collation:latin1_swedish_ci, user:root"] [command=Execute] [status="inTxn:1, autocommit:1"] [sql="UPDATE sbtest31 SET k=k+1 WHERE id=? [arguments: 35056900]"] [txn_mode=PESSIMISTIC] [timestamp=446735200326778960] [err="[tikv:9001]PD server timeout: \ngithub.com/pingcap/errors.AddStack\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20231212100244-799fae176cfb/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20231212100244-799fae176cfb/normalize.go:164\ngithub.com/pingcap/tidb/pkg/store/driver/error.ToTiDBErr\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/store/driver/error/error.go:112\ngithub.com/pingcap/tidb/pkg/store/driver/txn.extractKeyErr\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/store/driver/txn/error.go:164\ngithub.com/pingcap/tidb/pkg/store/driver/txn.(*tikvTxn).extractKeyErr\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/store/driver/txn/txn_driver.go:322\ngithub.com/pingcap/tidb/pkg/store/driver/txn.(*tikvTxn).LockKeysFunc\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/store/driver/txn/txn_driver.go:93\ngithub.com/pingcap/tidb/pkg/session.(*LazyTxn).LockKeysFunc\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/session/txn.go:483\ngithub.com/pingcap/tidb/pkg/session.(*LazyTxn).LockKeys\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/session/txn.go:458\ngithub.com/pingcap/tidb/pkg/executor.doLockKeys\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/executor.go:1279\ngithub.com/pingcap/tidb/pkg/executor.(*PointGetExecutor).lockKeyBase\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/point_get.go:409\ngithub.com/pingcap/tidb/pkg/executor.(*PointGetExecutor).lockKeyIfNeeded\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/point_get.go:384\ngithub.com/pingcap/tidb/pkg/executor.(*PointGetExecutor).getAndLock\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/point_get.go:369\ngithub.com/pingcap/tidb/pkg/executor.(*PointGetExecutor).Next\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/point_get.go:310\ngithub.com/pingcap/tidb/pkg/executor/internal/exec.Next\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/internal/exec/executor.go:309\ngithub.com/pingcap/tidb/pkg/executor.(*UpdateExec).updateRows\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/update.go:266\ngithub.com/pingcap/tidb/pkg/executor.(*UpdateExec).Next\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/update.go:240\ngithub.com/pingcap/tidb/pkg/executor/internal/exec.Next\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/internal/exec/executor.go:309\ngithub.com/pingcap/tidb/pkg/executor.(*ExecStmt).next\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/adapter.go:1245\ngithub.com/pingcap/tidb/pkg/executor.(*ExecStmt).handleNoDelayExecutor\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/adapter.go:990\ngithub.com/pingcap/tidb/pkg/executor.(*ExecStmt).handlePessimisticDML\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/adapter.go:1051\ngithub.com/pingcap/tidb/pkg/executor.(*ExecStmt).handleNoDelay\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/adapter.go:813\ngithub.com/pingcap/tidb/pkg/executor.(*ExecStmt).Exec\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/adapter.go:597\ngithub.com/pingcap/tidb/pkg/session.runStmt\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/session/session.go:2386\ngithub.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/session/session.go:2228\ngithub.com/pingcap/tidb/pkg/server.(*TiDBContext).ExecuteStmt\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/driver_tidb.go:293\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).executePreparedStmtAndWriteResult\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn_stmt.go:305\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).executePlanCacheStmt\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn_stmt.go:233\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).handleStmtExecute\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn_stmt.go:224\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).dispatch\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:1349\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).Run\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:1101\ngithub.com/pingcap/tidb/pkg/server.(*Server).onConn\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/server.go:701\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1650\nUPDATE sbtest31 SET k=k+1 WHERE id=? [arguments: 35056900]"]
[2024/01/02 09:41:57.282 +08:00] [INFO] [txn.go:55] ["Try to create a new txn inside a transaction auto commit"] [conn=3644851824] [session_alias=] [schemaVersion=55] [txnStartTS=446735200326778960] [txnScope=global]

// tidb-1
[2024/01/02 09:42:24.022 +08:00] [INFO] [domain.go:873] ["mdl gets lock, update to owner"] [jobID=415] [version=56]
[2024/01/02 09:42:24.092 +08:00] [INFO] [index.go:911] ["index backfill state ready to merge"] [category=ddl] ["job ID"=415] [table=sbtest1] [index=index_test_1704159486844]
[2024/01/02 09:42:24.225 +08:00] [INFO] [domain.go:873] ["mdl gets lock, update to owner"] [jobID=415] [version=57]
[2024/01/02 09:42:24.318 +08:00] [INFO] [index.go:1959] ["start to merge temp index"] [category=ddl] [job="ID:415, Type:add index, State:running, SchemaState:write reorganization, SchemaID:102, TableID:245, RowCount:70000034, ArgLen:6, start time: 2024-01-02 09:38:06.845 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:446735251995361348, UniqueWarnings:0"] [reorgInfo="CurrElementType:_idx_,CurrElementID:2,StartKey:7480000000000000f55f69ffff000000000002,EndKey:7480000000000000f55f69ffff000000000002ff,First:false,PhysicalTableID:245,Ingest mode:false"]
[2024/01/02 09:42:25.610 +08:00] [INFO] [index.go:742] ["run add index job done"] [category=ddl] [charset=utf8mb4] [collation=utf8mb4_general_ci]
[2024/01/02 09:42:25.653 +08:00] [INFO] [domain.go:873] ["mdl gets lock, update to owner"] [jobID=415] [version=58]

// tidb-0
[2024/01/02 09:42:27.559 +08:00] [INFO] [domain.go:915] ["reload schema in loop, schema syncer need restart"]
[2024/01/02 09:42:27.559 +08:00] [INFO] [schema_validator.go:98] ["the schema validator stops"]
[2024/01/02 09:42:27.573 +08:00] [INFO] [schema_validator.go:129] ["the schema validator stopped before updating"]
[2024/01/02 09:42:27.573 +08:00] [INFO] [domain.go:973] ["mustReload succeed"]
[2024/01/02 09:42:27.573 +08:00] [INFO] [schema_validator.go:109] ["the schema validator restarts"]
[2024/01/02 09:42:27.573 +08:00] [INFO] [domain.go:936] ["schema syncer restarted"]
[2024/01/02 09:42:27.573 +08:00] [INFO] [domain.go:272] ["diff load InfoSchema success"] [currentSchemaVersion=55] [neededSchemaVersion=58] ["start time"=7.716326ms] [gotSchemaVersion=58] [phyTblIDs="[245,245,245]"] [actionTypes="[7,7,7]"] [diffTypes="[\"add index\",\"add index\",\"add index\"]"]
[2024/01/02 09:42:30.123 +08:00] [INFO] [ddl.go:1165] ["DDL job is finished"] [category=ddl] [jobID=415]
```

```
// tidb-0
[2024/01/02 09:42:30.135 +08:00] [INFO] [session.go:3925] [GENERAL_LOG] [conn=3644851796] [session_alias=] [user=root@10.233.71.65] [schemaVersion=58] [txnStartTS=446735253528903767] [forUpdateTS=446735253528903767] [isReadConsistency=false] [currentDB=sysbench_64_7000w] [isPessimistic=true] [sessionTxnMode=PESSIMISTIC] [sql="SELECT SUM(k) FROM sbtest1 WHERE id BETWEEN ? AND ? [arguments: (34975499, 34975598)] [arguments: (34975499, 34975598)]"]
[2024/01/02 09:42:30.135 +08:00] [INFO] [session.go:3925] [GENERAL_LOG] [conn=3644851824] [session_alias=] [user=root@10.233.71.65] [schemaVersion=55] [txnStartTS=446735253528903726] [forUpdateTS=446735253528903726][isReadConsistency=false] [currentDB=sysbench_64_7000w] [isPessimistic=true] [sessionTxnMode=PESSIMISTIC] [sql="DELETE FROM sbtest1 WHERE id=? [arguments: 45356626] [arguments: 45356626]"]
```

The start ts of two DML txns is close, but the schema version differs(55 v.s. 58). The txn with schemaVersion = 55 first obtains the info schema from the cache before the restoration of network partition (that is, the domain has not been successfully reloaded). When it commit later, because MDL is enabled, the schema validator check will be skipped, resulting in the commit of data under an old schema (index state is `write-reorg`, backfilling state is `BackfillStateRunning`, the write is redirected to temporary index), which further leads to inconsistent data indexes.

Here is the timeline:

1. Add index to tidb-0, change index status to delete-only.
2. Network partitioned between tidb-0 and PD, and tidb-1 is elected as the DDL owner.
3. tidb-1 continues to add indexes until public, schema version advances from 55 to 58.
4. tidb-0 network partition disappears, schema version is still 55.
5. tidb-0 executes DML using schema version 55 and commits.

### What changed and how does it work?

- When tidb-0 comes back, record the latest schema version in cache.
- Check version in schema validator, return an error if a txn's schema version is too small.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/domain/schema_validator.go [hunk 1]
```diff
@@ -228,6 +236,11 @@ func (s *schemaValidator) Check(txnTS uint64, schemaVer int64, relatedPhysicalTa
 		logutil.BgLogger().Info("the schema validator stopped before checking")
 		return nil, ResultUnknown
 	}
+
+	if schemaVer < s.restartSchemaVer {
+		logutil.BgLogger().Info("the schema version is too old", zap.Int64("schemaVer", schemaVer))
```
Comment (wjhuang2016) [inline pkg/domain/schema_validator.go:241]: add `please check if the tidb and pd are healthy after the transaction started`

**Review by wjhuang2016 — APPROVED** [review]

**Review by zimulala — APPROVED** [review]

**Pull request merged**